src/binary_basic.ml

```diff
@@ -0,0 +1,534 @@
+open Types
+open Syntax
+
+module Input = struct
+  type t =
+    { bytes : bytes
+    ; pt : int
+    ; size : int
+    }
+
+  let _pp input =
+    print_int input.pt;
+    print_string " ";
+    print_int input.size;
+    print_newline ()
+
+  let is_empty input = input.size = 0
+
+  let from_bytes bytes =
+    let size = Bytes.length bytes in
+    { bytes; pt = 0; size }
+
+  let sub ~pos ~len input =
+    if pos <= input.size && len <= input.size - pos then
+      Ok { input with pt = input.pt + pos; size = len }
+    else Error (`Msg "Input.sub error")
+
+  let sub_suffix pos input = sub ~pos ~len:(input.size - pos) input
+
+  let sub_prefix len input = sub ~pos:0 ~len input
+
+  let get n input =
+    if n < input.size then Ok (Bytes.get input.bytes (input.pt + n))
+    else Error (`Msg "Input.get error")
+
+  let get0 = get 0
+end
+
+let string_of_char_list char_list =
+  let buf = Buffer.create 64 in
+  List.iter (Buffer.add_char buf) char_list;
+  Buffer.contents buf
+
+let read_byte input =
+  let* c = Input.get0 input in
+  let* next_input = Input.sub_suffix 1 input in
+  Ok (c, next_input)
+
+(* https://en.wikipedia.org/wiki/LEB128#Unsigned_LEB128 *)
+let rec read_UN n input =
+  let* b, input = read_byte input in
+  let b = Char.code b in
+  let x = Int64.of_int (b land 0x7f) in
+  if b land 0x80 = 0 then Ok (x, input)
+  else
+    let* i64, input = read_UN (n - 7) input in
+    Ok (Int64.logor x (Int64.shl i64 7L), input)
```

Comment:
Depending on our perf requirements, we may want to make this tail-rec

Reply:
Add TODO